parsers/ES.py

```diff
@@ -30,91 +28,109 @@
-ZONE_KEYS = Literal[
-    "ES",  # Spain
-    "ES-CE",  # Ceuta
-    "ES-CN-FVLZ",  # Fuerteventura/Lanzarote
-    "ES-CN-GC",  # Gran Canaria
-    "ES-CN-HI",  # El Hierro
-    "ES-CN-IG",  # Isla de la Gomera
-    "ES-CN-LP",  # La Palma
-    "ES-CN-TE",  # Tenerife
-    "ES-IB",  # Balearic Islands
-    "ES-IB-FO",  # Formentera
-    "ES-IB-IZ",  # Ibiza
-    "ES-IB-MA",  # Mallorca
-    "ES-IB-ME",  # Menorca
-    "ES-ML",  # Melilla
-]
-
-# TODO: Update floors to be non zero.
-# Minimum valid zone demand. This is used to eliminate some cases
-# where generation for one or more modes is obviously missing.
-ZONE_FLOORS: Dict[ZONE_KEYS, float] = {
-    "ES": 0,
-    "ES-CE": 0,
-    "ES-CN-FVLZ": 50,
-    "ES-CN-GC": 150,
-    "ES-CN-HI": 2,
-    "ES-CN-IG": 3,
-    "ES-CN-LP": 10,
-    "ES-CN-TE": 150,
-    "ES-IB": 0,
-    "ES-IB-FO": 0,
-    "ES-IB-IZ": 0,
-    "ES-IB-MA": 0,
-    "ES-IB-ME": 0,
-    "ES-ML": 0,
-}
 
-ZONE_FUNCTION_MAP: Dict[ZONE_KEYS, Callable] = {
-    "ES": IberianPeninsula,
-    "ES-CE": Ceuta,
-    "ES-CN-FVLZ": LanzaroteFuerteventura,
-    "ES-CN-GC": GranCanaria,
-    "ES-CN-HI": ElHierro,
-    "ES-CN-IG": Gomera,
-    "ES-CN-LP": LaPalma,
-    "ES-CN-TE": Tenerife,
-    "ES-IB": BalearicIslands,
-    "ES-IB-FO": Formentera,
-    "ES-IB-IZ": Ibiza,
-    "ES-IB-MA": Mallorca,
-    "ES-IB-ME": Menorca,
-    "ES-ML": Melilla,
+ZONE_FUNCTION_MAP: Dict[ZoneKey, Callable] = {
+    ZoneKey("ES"): IberianPeninsula,
+    ZoneKey("ES-CE"): Ceuta,
+    ZoneKey("ES-CN-FVLZ"): LanzaroteFuerteventura,
+    ZoneKey("ES-CN-GC"): GranCanaria,
+    ZoneKey("ES-CN-HI"): ElHierro,
+    ZoneKey("ES-CN-IG"): Gomera,
+    ZoneKey("ES-CN-LP"): LaPalma,
+    ZoneKey("ES-CN-TE"): Tenerife,
+    ZoneKey("ES-IB-FO"): Formentera,
+    ZoneKey("ES-IB-IZ"): Ibiza,
+    ZoneKey("ES-IB-MA"): Mallorca,
+    ZoneKey("ES-IB-ME"): Menorca,
+    ZoneKey("ES-ML"): Melilla,
 }
 
 EXCHANGE_FUNCTION_MAP: Dict[str, Callable] = {
-    "ES->ES-IB": BalearicIslands,
     "ES->ES-IB-MA": Mallorca,
     "ES-IB-IZ->ES-IB-MA": Mallorca,
     "ES-IB-FO->ES-IB-IZ": Formentera,
     "ES-IB-MA->ES-IB-ME": Mallorca,
 }
 
+# dict of REE production types to Electricity Maps production types
+PRODUCTION_MAPPING = {
+    "carbon": "coal",
+    "combined": "gas",
+    "diesel": "oil",
+    "gas": "gas",
+    "hydraulic": "hydro",
+    "nuclear": "nuclear",
+    "other": "unknown",
+    "solar": "solar",
+    "vapor": "oil",
+    "waste": "biomass",
+    "wind": "wind",
+}
 
-def fetch_island_data(
-    zone_key: ZONE_KEYS, session: Session, target_datetime: Optional[datetime]
-) -> List[Response]:
-    if isinstance(target_datetime, datetime):
-        date = target_datetime.strftime("%Y-%m-%d")
+
+def check_valid_parameters(
+    zone_key: ZoneKey,
+    session: Optional[Session],
+    target_datetime: Optional[datetime],
+    logger: Logger,
+):
+    """Raise an exception if the parameters are not valid for this parser."""
+    if "->" in zone_key:
```

Comment:
Hmm, while this might make readability easier won't it be a performance hit?

We would have to import EXCHANGES_CONFIG which in turn needs to read the files. It would also make us miss edge cases if the exchange don't exist at all due to a miss spelling or errors in parameter sorting/passing.

The current way should be very performant with just one logic check then branching after that with 1 logic check.

The your option would also have 2 logic checks but it would require additional imports and file reading as well?

I think we should prioritize performance in this parser as it's run often and even if I added the check it is improbable we will hit it.

Reply:
Pushed a change that I think is the best of both worlds, let me know what you think.